privacy-scaling-explorations/e2e-zk-ecdsa
# ✨ Implement message queue to process proofs requests

Closes #114 

# Description
The proof generation is a CPU intensive process that blocks the main thread/event loop.
It is problem because while the proof is being is being generate (the `snarkjs` command), the server can't not do anything else.
- it can't answer the request it has received. The http request connection with the client needs to stay opened for 5 -10 min which is bad UX for the client. Moreover node fetch times out after 3min anyway.
- it can't receive other requests
- it is a security risk as it becomes very easy to DoS the server

Instead we should use a message queue so that we split the tasks of handling the requests and processing them into different processes and threads.
This PR introduces a redis based message queue ([bullmq](https://bullmq.io/)) for that purpose

## Behavior Changes

| Before | After |
| ------ | ----- |
|  Send proof request to `/` <br>Wait 8min for completion (http connection still open). <br>Get the result back as json     |  Send proof request to `/proof`. <br>Get immediately a job id back, http connection closes.<br>Wait 8 min<br>Make a new request to `/proofs/:id/(proof,public).json` to get proof files served statically     |

# How has this been tested?
1. In root: build shared and proving-api packages: `lerna run build --scope '{@e2e-zk-ecdsa/proving-api,@e2e-zk-ecdsa/shared}'`
2. In `apis/proving`:
  1. Have redis running: `docker compose up`
  2. start server: `pnpm run server.prod`
  3. make request: `pnpm run client.prod`

The client should get a response like:
```
 Proof Generation Job b30bf08550296438a3b3cf19d769d10bed9763e1b4fb12299b91ca39d445f027 added to queue.
    It might take 5- 10 min to complete.
    Check your results later at
    /proofs/b30bf08550296438a3b3cf19d769d10bed9763e1b4fb12299b91ca39d445f027/(input.json|proof.json|public.json|witness.wtns).
```

Some stuff should be logged server side too.

In the browser you can check on the queue dashboard the status of the job at http://localhost:3000/dashboard/:
<img width="1248" alt="Screenshot 2023-04-05 at 09 41 44" src="https://user-images.githubusercontent.com/38692952/230014348-00e501eb-2fea-4709-918b-9f33312c8bfa.png">

Once the job has completed, you can get the proof results/artifacts at
`http://localhost:3000/proofs/:id/proof.json`
`http://localhost:3000/proofs/:id/public.json`
`http://localhost:3000/proofs/:id/input.json`
`http://localhost:3000/proofs/:id/witness.wtns`

## TODO
Right now the only protection against someone having access to someone else's zk proofs is the generated id. Right now it is only done by hashing the request and adding some salt. We should implement something more robust later (JWT authentication?) see #121.
We can also add another queue/worker for expiration: e.g. automatically remove the static files once they were fetched or after some time. see #120

# Checklist

- [ ] All existing CI checks pass
  - My code compiles
  - My code follows the style guidelines of this project (code is linted and formatted)
  - My code doesn't break existing tests
- [x] I added a reviewer for this pull request.
- [x] I have added the appropriate label(s) to this pull request.
